src/solver/opamBuiltin0install.ml.real

```diff
@@ -0,0 +1,99 @@
+}
+
+let not_relop = function
+  | `Eq -> `Neq
+  | `Neq -> `Eq
+  | `Geq -> `Lt
+  | `Gt -> `Leq
+  | `Leq -> `Gt
+  | `Lt -> `Geq
+
+let keep_installed request pkgname =
+  not (List.exists (fun (pkg, _) -> String.equal pkg pkgname) request.Cudf.install) &&
+  not (List.exists (fun (pkg, _) -> String.equal pkg pkgname) request.Cudf.upgrade) &&
+  not (List.exists (fun (pkg, _) -> String.equal pkg pkgname) request.Cudf.remove)
+
+let add_spec pkg req c (pkgs, constraints) =
+  let pkgs = (pkg, req) :: pkgs in
+  let constraints = match c with
+    | None -> constraints
+    | Some c -> (pkg, c) :: constraints
+  in
+  (pkgs, constraints)
+
+let essential spec (pkg, c) = add_spec pkg `Essential c spec
+let recommended spec (pkg, c) = add_spec pkg `Recommended c spec
+
+let restricts (pkgs, constraints) (pkg, c) =
+  let constraints = match c with
+    | None -> (pkg, (`Lt, 1)) :: (pkg, (`Gt, 1)) :: constraints (* pkg < 1 & pkg > 1 is always false *)
+    | Some (relop, v) -> (pkg, (not_relop relop, v)) :: constraints
+  in
+  (pkgs, constraints)
+
+let create_spec universe request =
+  let spec = ([], []) in
+  let spec = List.fold_left essential spec request.Cudf.install in
+  let spec = List.fold_left essential spec request.Cudf.upgrade in
+  let spec = List.fold_left restricts spec request.Cudf.remove in
+  Cudf.fold_packages_by_name (fun spec pkgname pkgs ->
+      match List.find_opt (fun pkg -> pkg.Cudf.installed) pkgs with
+      | Some {Cudf.keep = `Keep_version; version; _} -> essential spec (pkgname, Some (`Eq, version))
+      | Some {Cudf.keep = `Keep_package; _} -> essential spec (pkgname, None)
+      | Some {Cudf.keep = `Keep_feature; _} -> assert false (* TODO: Is this used? *)
+      | Some {Cudf.keep = `Keep_none; _} when keep_installed request pkgname -> recommended spec (pkgname, None)
+      | Some {Cudf.keep = `Keep_none; _}
+      | None -> spec
+    ) spec universe
+
+let reconstruct_universe universe selections =
+  Opam_0install_cudf.packages_of_result selections |>
+  List.fold_left (fun pkgs (pkg, v) ->
+      let pkg = Cudf.lookup_package universe (pkg, v) in
+      {pkg with was_installed = pkg.installed; installed = true} :: pkgs
+    ) [] |>
+  Cudf.load_universe
+
+let call ~criteria:_ ?timeout:_ (preamble, universe, request) =
+  let t0 = Unix.gettimeofday () in
+  let pkgs, constraints = create_spec universe request in
+  let context = Opam_0install_cudf.create ~constraints universe in
+  match Opam_0install_cudf.solve context pkgs with
+  | Ok selections ->
+    let universe = reconstruct_universe universe selections in
+    let t1 = Unix.gettimeofday () in
+    log "Solution found. Solve took %.2f s" (t1 -. t0);
+    (Some preamble, universe)
+  | Error _problem ->
+    let t1 = Unix.gettimeofday () in
+    log "No solution. Solve took %.2f s" (t1 -. t0);
+    (* Diagnostics disabled. Feel free to enable again for debug *)
+    (* print_endline (Opam_0install_cudf.diagnostics problem); *)
```

Comment:
`OpamConsole.log ~level:3 "%a" (slog Opam_0install_cudf.diagnostics) problem` ?

Reply:
Oh nice, I hadn't thought of that. Done.